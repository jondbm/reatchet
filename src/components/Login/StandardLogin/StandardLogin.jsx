import React, { Component } from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'

import TextInput from '../../common/TextInput'
import Button from '../../common/Buttons/Button'
import { Wrapper, Subtitle, StyledRecaptcha } from '../sharedComponents'

import { validateEmail } from '../../../helpers/validation'
import theme from '../../../theme'

/*
*****************
STYLED COMPONENTS
*****************
*/

const Error = styled.span`
  color: ${theme.colours.redDark};
  text-align: left;
  align-self: flex-start;
  margin-bottom: 12px;
  font-size: 12px;
`

const LoginButton = styled(Button)`
  margin-top: ${theme.spacing.sm};
  align-self: stretch;
`

const ForgottenPasswordLink = styled(Link)`
  font-size: 12px;
  color: ${theme.colours.teal};
  text-align: center;
  align-self: center;
  margin-top: ${theme.spacing.xs};
  text-decoration: none;
`

/*
*********
COMPONENT
*********
*/

class StandardLogin extends Component {
  state = {
    email: '',
    password: '',
    validation: {
      email: false,
      password: false
    },
    recaptchaToken: ''
  }

  updateEmail = value => {
    const isValid = validateEmail(value)

    this.setState({
      email: value,
      validation: {
        ...this.state.validation,
        email: isValid
      }
    })
  }

  updatePassword = value => {
    const isValid = value.length >= 8

    this.setState({
      password: value,
      validation: {
        ...this.state.validation,
        password: isValid
      }
    })
  }

  handleRecaptcha = token => {
    this.setState({ recaptchaToken: token })
  }

  handleExpiredRecaptcha = () => {
    this.setState({ recaptchaToken: '' })
  }

  logIn = () => {
    const { email, password } = this.state
    this.props.logIn(email, password)
  }

  // TODO: re-enable production check for recaptcha value once k8 is setup to handle it
  isButtonEnabled = () => {
    // if (process.env.REACT_APP_SEENIT_ENV === 'production') {
    //   return (
    //     this.state.recaptchaToken &&
    //     Object.values(this.state.validation).every(value => !!value)
    //   )
    // }
    return Object.values(this.state.validation).every(value => !!value)
  }

  render() {
    const buttonIsDisabled = !this.isButtonEnabled()

    return (
      <Wrapper>
        <Subtitle>
          Log in to Seenit Studio with your email address and password
        </Subtitle>
        <TextInput
          label="Email address"
          handleChange={this.updateEmail}
          value={this.state.email}
          isValid={this.state.validation.email}
          validationMessage="Enter a valid email address"
          data-key="input-email"
        />

        <TextInput
          label="Password"
          isPasswordInput
          handleChange={this.updatePassword}
          value={this.state.password}
          isValid={this.state.validation.password}
          validationMessage="Passwords must be at least 8 characters long"
          data-key="input-password"
        />

        {this.props.hasInvalidCredentials && (
          <Error data-el="credentials-error">
            Oops! Invalid email address or password. Please try again.
          </Error>
        )}

        <StyledRecaptcha
          handleRecaptcha={this.handleRecaptcha}
          handleExpiredRecaptcha={this.handleExpiredRecaptcha}
          id="login"
        />
        <LoginButton
          data-el="login-button"
          variant="raised"
          onClick={this.logIn}
          disabled={buttonIsDisabled}
        >
          Log in
        </LoginButton>
        <ForgottenPasswordLink
          to="/login/forgotten-password"
          data-el="forgotten-password-link"
        >
          Forgotten your password?
        </ForgottenPasswordLink>
      </Wrapper>
    )
  }
}

export default StandardLogin
