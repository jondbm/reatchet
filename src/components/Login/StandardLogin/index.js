import { connect } from 'react-redux'

import StandardLogin from './StandardLogin'
import * as actions from '../../../reducers/user/actions'

export const mapStateToProps = state => ({
  hasInvalidCredentials: state.user.hasInvalidCredentials
})

export default connect(mapStateToProps, { ...actions })(StandardLogin)
