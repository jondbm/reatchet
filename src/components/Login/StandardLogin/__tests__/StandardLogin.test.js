import React from 'react'
import 'jest-styled-components'

import { shallow } from 'enzyme'

import StandardLogin from '../StandardLogin'

describe('<StandardLogin /> component', () => {
  const initialState = {
    email: '',
    password: '',
    validation: {
      email: false,
      password: false
    },
    recaptchaToken: ''
  }

  let component
  const logIn = jest.fn()

  const props = {
    logIn
  }

  beforeEach(() => {
    component = shallow(<StandardLogin {...props} />)
  })

  describe('basic rendering', () => {
    it('should correctly render the component', () => {
      expect(component.exists()).toBe(true)
    })

    it('should match the snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })

  describe('component methods', () => {
    test('updateEmail', () => {
      const email = 'test@test.com'
      const expectedValidationState = {
        ...initialState.validation,
        email: true
      }

      const updateEmail = component.instance().updateEmail
      updateEmail(email)
      component.update()
      expect(component.state('email')).toEqual(email)
      expect(component.state('validation')).toEqual(expectedValidationState)
    })

    test('updatePassword', () => {
      const password = 'paswword123'
      const expectedValidationState = {
        ...initialState.validation,
        password: true
      }

      const updatePassword = component.instance().updatePassword
      updatePassword(password)
      component.update()
      expect(component.state('password')).toEqual(password)
      expect(component.state('validation')).toEqual(expectedValidationState)
    })

    test('handleRecaptcha', () => {
      const handleRecaptcha = component.instance().handleRecaptcha
      handleRecaptcha('123456')
      component.update()
      expect(component.state('recaptchaToken')).toEqual('123456')
    })

    test('handleExpiredRecaptcha', () => {
      const handleExpiredRecaptcha = component.instance().handleExpiredRecaptcha
      handleExpiredRecaptcha()
      component.update()
      expect(component.state('recaptchaToken')).toEqual('')
    })

    test('logIn', () => {
      const instanceLogIn = component.instance().logIn
      const email = 'test@test.com'
      const password = 'password123'
      component.setState({ email, password })
      component.update()
      instanceLogIn()
      component.update()
      expect(logIn).toHaveBeenCalledWith(email, password)
    })
  })
})
