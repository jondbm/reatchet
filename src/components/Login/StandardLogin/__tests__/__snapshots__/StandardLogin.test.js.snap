// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StandardLogin /> component basic rendering should match the snapshot 1`] = `
<sharedComponents__Wrapper>
  <sharedComponents__Subtitle>
    Log in to Seenit Studio with your email address and password
  </sharedComponents__Subtitle>
  <TextInput
    data-key="input-email"
    handleChange={[Function]}
    isPasswordInput={false}
    isValid={false}
    label="Email address"
    validationMessage="Enter a valid email address"
    value=""
  />
  <TextInput
    data-key="input-password"
    handleChange={[Function]}
    isPasswordInput={true}
    isValid={false}
    label="Password"
    validationMessage="Passwords must be at least 8 characters long"
    value=""
  />
  <sharedComponents__StyledRecaptcha
    handleExpiredRecaptcha={[Function]}
    handleRecaptcha={[Function]}
    id="login"
  />
  <StandardLogin__LoginButton
    data-el="login-button"
    disabled={true}
    onClick={[Function]}
    variant="raised"
  >
    Log in
  </StandardLogin__LoginButton>
  <StandardLogin__ForgottenPasswordLink
    data-el="forgotten-password-link"
    to="/login/forgotten-password"
  >
    Forgotten your password?
  </StandardLogin__ForgottenPasswordLink>
</sharedComponents__Wrapper>
`;
